test: update test_parse_type() regex

helloibis opened the pull request:
### What I did
The previous regex was allowing some invalid inputs such as `(,)` to get through.

### How I did it
Updated the regular expression to match expected Vyper decls like:
- type identifiers
- array declarations
- tuples of type/array
- tuples of type/array/tuple/etc

I split the various expected sub-patterns to their own variables so they're easier to read and can be commented on separately

### How to verify it

__NOTE:__ I used non-capturing groups (?:) on the regex site just to make the highlight colors a little less chaotic 👀 the groups in the Python code don't need to be non-capturing so I omitted them there

The regular expression (i.e. the value of `TYPE_DECL`) and some sample input are here:

https://regex101.com/r/dVmiYA/1

Screenshot for posterity, the non-highlighted text below shows invalid input that we don't want to use anyway:

<img width="703" alt="image" src="https://user-images.githubusercontent.com/108302637/183513383-798c37f0-6edb-4f07-afcf-f05d631c913b.png">

### Checklist
- [x] All changes are completed
- [x] New test cases have been added
- [x] Documentation has been updated

unparalleled-js reviewed (approved):
Looks great!

NotPeopling2day commented on tests/functional/utils/test_abi.py:
```diff
@@ -3,18 +3,46 @@
+    #   - (int[],)
+    #   - (int, int,)
+    #   - (int, int)
+    #   - (int, int, int,)
+    #   - (int, int, int,)
```
Are these duplicates, or maybe missed removing the comma?

helloibis replied:
Updated in 1b16442c

helloibis commented:
Updated 54a8bc14 just because we don't need a separate `NESTED_TUPLE` pattern, just need to redefine `TUPLE` and nest its existing value inside of it. That helps reduce the total length of the regex which should give hypothesis a little bit of a break lol

@NotPeopling2day I fixed the docstring in 1b16442c

I also updated the description to reflect the consolidated regex

NotPeopling2day reviewed (dismissed):
LGTM
